lib/src/encoder/extensions.dart

```diff
@@ -0,0 +1,59 @@
+part of 'encoder.dart';
+
+extension _AccountsIterableExt on Iterable<AccountMeta> {
+  // Convert account metas to encoder public keys
+  Iterable<Buffer> toSerializablePubKeys() =>
+      map((a) => Buffer.fromBase58(a.pubKey));
+}
+
+extension _InstructionsIterableExt on Iterable<Instruction> {
+  /// Simple helper to extract all accounts from an instruction, it also appends
+  /// the program id account which needs to be present in the accounts array sent
+  /// in a message.
+  Iterable<AccountMeta> _extractAccounts(String? feePayer) {
+    final accounts = [
+      for (final Instruction instruction in this) ...[
+        ...instruction.accounts,
+      ],
+      for (final Instruction instruction in this) ...[
+        AccountMeta.readonly(pubKey: instruction.programId, isSigner: false),
+      ]
+    ];
+    if (feePayer == null) {
+      return accounts;
+    }
+    final feePayerIndex = accounts.indexOfPubKey(feePayer);
+    if (feePayerIndex == -1) {
+      return [
+        AccountMeta.writeable(pubKey: feePayer, isSigner: true),
+        ...accounts,
+      ];
+    } else {
+      return [
+        AccountMeta.writeable(
+          pubKey: feePayer,
+          isSigner: true,
+        ),
+        ...accounts.sublist(0, feePayerIndex),
+        ...accounts.sublist(feePayerIndex + 1),
+      ];
+    }
+  }
+
+  /// Combines accounts from every instruction and sorts them according
+  /// to [Account Addresses Format][account addresses format], also removes
+  /// any duplicates.
+  ///
+  /// Duplicates are removed by picking signers and writeable accounts
+  /// over non-writeable and non-signers.
+  ///
+  /// Accounts are considered duplicates if they have the same public keys.
+  ///
+  /// [account addresses format]: https://docs.solana.com/developing/programming-model/transactions#account-addresses-format
+  Iterable<AccountMeta> _getAccounts(String? feePayer) =>
```

Comment:
If I got correctly what you mean here, I suggest changing this file to the following (and removing obsolete code from other files):

```dart
part of 'encoder.dart';

extension on Iterable<AccountMeta> {
  // Convert account metas to encoder public keys
  Iterable<Buffer> toSerializablePubKeys() =>
      map((a) => Buffer.fromBase58(a.pubKey));

  /// Removes all duplicate accounts, meaning accounts having the same public
  /// key. Signers and/or writeable accounts are picked over non-signers and
  /// non-writeable.
  List<AccountMeta> unique() =>
      fold<List<AccountMeta>>([], (List<AccountMeta> list, AccountMeta item) {
        final index = list.indexWhere((a) => a.pubKey == item.pubKey);
        if (index == -1) {
          list.add(item);
        } else {
          // Keep then one of the two that is either a signer or
          // a writeable account
          list[index] = item.mergeWith(list[index]);
        }

        return list;
      });
}

extension on Iterable<Instruction> {
  /// Prepares all the accounts from instructions:
  ///
  /// - extracts accounts from every instruction;
  /// - appends program id accounts (accounts with the sample public keys);
  /// - removes duplicates (by picking signers and writeable accounts
  ///   over non-writeable and non-signers);
  /// - sorts accounts according to [Account Addresses Format][1].
  ///
  /// [1]: https://docs.solana.com/developing/programming-model/transactions#account-addresses-format
  Iterable<AccountMeta> getAccounts(String? feePayer) {
    final accounts = expand((i) => [
          ...i.accounts,
          AccountMeta.readonly(pubKey: i.programId, isSigner: false),
        ]);

    return [
      ...accounts,
      if (feePayer != null)
        AccountMeta.writeable(pubKey: feePayer, isSigner: true),
    ].unique()
      ..sort(_accountComparator);
  }
}

extension on Blockhash {
  ByteArray toBytes() => Buffer.fromBase58(blockhash);
}

/// Compare accounts according to the following rules
///
/// Signer accounts go first, and within them writeable accounts
/// go first
///
/// Non-Signer accounts go after, and within them the writeable
/// accounts go first
int _accountComparator(AccountMeta a1, AccountMeta a2) {
  if (a1.isSigner && !a2.isSigner) return -1;
  if (!a1.isSigner && a2.isSigner) return 1;

  if (a1.isWriteable && !a2.isWriteable) return -1;
  if (!a1.isWriteable && a2.isWriteable) return 1;

  return 0;
}
```

I think it will make it more readable.